File: 0768-max-chunks-to-make-sorted-ii/0768-max-chunks-to-make-sorted-ii.java
class Solution {
    public int maxChunksToSorted(int[] arr) {
        
        // pre compute min array from right to left
        int[] minArray = new int[arr.length];
        int n = arr.length;
        minArray[n-1] = arr[n-1];
        for(int i = n-2; i>=0 ;i--){
            minArray[i] = Math.min(minArray[i+1], arr[i]);
        }
        
        // calculate max from left to right (no need to store as we just need curr max)
        int max = Integer.MIN_VALUE;
        int count = 0;
        for(int i = 0; i < n; i++){
            if(max < arr[i]){
                max = arr[i];
            }
            // if current max <= next min -> can chunk
            if(i+1 < n && max <= minArray[i+1])
                count++;
        }
        return count+1; // add 1 for last chunk
    }
}

File: 0768-max-chunks-to-make-sorted-ii/0768-max-chunks-to-make-sorted-ii.kt
class Solution {
    fun maxChunksToSorted(arr: IntArray): Int {
        
        
        // pre computation approach
        val minArr = IntArray(arr.size) { ind -> arr[ind] }
        
        for(ind in arr.size-2 downTo 0){
            minArr[ind] = Math.min(minArr[ind+1], minArr[ind])
        }
        var max = Int.MIN_VALUE // max until now
        var count = 0
        for(i in arr.indices){
            if(arr[i] > max)
                max = arr[i]
            if(i+1< arr.size && max <= minArr[i+1]){ // max until now <= min from next
                count++
            }
        }
        return count+1  
    }
}